### PR TITLE
[HTTP/2] Throw meaningful exception if we get GOAWAY while reading response body

### LINKED ISSUE #97128: HTTP/2 request doesn't report the correct reason why request ends on connection GOAWAY
### Description

A customer ran into a HTTP/2 error scenario and asked why the real failure reason wasn't in the error message: https://github.com/grpc/grpc-dotnet/issues/2361#issuecomment-1895805167. Seeing this error is important for debugging because the server can trigger a GOAWAY and knowing the error code is important for understanding why the server closed the connection.

Scenario:
1. Response data from a HTTP/2 request is being read from the response stream
2. The underlying HTTP/2 connection is killed because of a GOAWAY form the server
3. The user sees an unhelpful/deceptive message:

```
System.Net.Http.HttpIOException: The response ended prematurely while waiting for the next frame from the server. (ResponseEnded)
   at System.Net.Http.Http2Connection.ThrowRequestAborted(Exception innerException)
   at System.Net.Http.Http2Connection.Http2Stream.CheckResponseBodyState()
   at System.Net.Http.Http2Connection.Http2Stream.TryReadFromBuffer(Span`1 buffer, Boolean partOfSyncRead)
   at System.Net.Http.Http2Connection.Http2Stream.ReadDataAsync(Memory`1 buffer, HttpResponseMessage responseMessage, CancellationToken cancellationToken)
```

The error gives the impression that the request ended because the server unexpectedly stopped sending response data. The real reason there are no more frames is because the server closed the connection with a GOAWAY. That information (GOAWAY + enhance your calm protocol code) should be bubbled up to this error message.

Something like:

> The response ended because the server ended the HTTP/2 connection. Error code: ENHANCE_YOUR_CALM.

Today the only way to figure out why the HTTP/2 request failed is to attach an event source listener. That displays:

```
ProcessIncomingFramesAsync,Frame 15: StreamId=0; Type=GoAway; Flags=None; PayloadLength=22.
ReadGoAwayFrame,lastStreamId=1, errorCode=EnhanceYourCalm
```

Setting up an event source listener is advanced and shouldn't be required to find out important error information like this.

### Reproduction Steps

1. Client starts an HTTP/2 request and awaits response headers
4. Server returns response headers
5. Client reads from response stream
6. Server sends a GOAWAY
7. Client gets the unhelpful error

Here is an example in logs:

[ping-goaway.txt](https://github.com/dotnet/runtime/files/13971130/ping-goaway.txt)

### Expected behavior

A more useful error message. It should say the response ended because the connection ended. It should include the GOAWAY error code.

Something like:

> The response ended because the server ended the HTTP/2 connection. Error code: ENHANCE_YOUR_CALM.

### Actual behavior

Unhelpful error message.

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

_No response_

### Other information

_No response_

### PR DESCRIPTION
**Pull request by liveans**
Fixes #97128

Not sure if this is the correct way to fix this.

### CODE REVIEW

File: src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http2.cs [hunk 1]
```diff
@@ -1051,6 +1051,39 @@ public async Task GoAwayFrame_RequestWithBody_ServerDisconnect_AbortStreamsAndTh
             }
         }
 
+        [ConditionalFact(nameof(SupportsAlpn))]
+        public async Task GoAwayFrame_RequestServerDisconnects_ThrowsHttpProtocolExceptionWithProperErrorCode()
+        {
+            _output.WriteLine("Started!");
```
Comment (MihaZupan) [inline src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http2.cs:1057]:
```suggestion
```

File: src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http2.cs [hunk 2]
```diff
@@ -1051,6 +1051,39 @@ public async Task GoAwayFrame_RequestWithBody_ServerDisconnect_AbortStreamsAndTh
...
+                HttpProtocolException exception = await Assert.ThrowsAsync<HttpProtocolException>(() => responseStream.ReadAsync(buffer).AsTask());
+                Assert.Equal(ProtocolErrors.ENHANCE_YOUR_CALM, (ProtocolErrors) exception.ErrorCode);
+                Assert.IsType<HttpIOException>(exception.InnerException);
+                Assert.Contains("ended prematurely while waiting", exception.Message);
+
+            },
+            async server =>
+            {
+                // Server returns response headers
+                Http2LoopbackConnection connection = await server.EstablishConnectionAsync();
```
Comment (MihaZupan) [inline src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http2.cs:1076]:
```suggestion
                await using Http2LoopbackConnection connection = await server.EstablishConnectionAsync();
```

File: src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http2.cs [hunk 3]
```diff
@@ -1051,6 +1051,39 @@ public async Task GoAwayFrame_RequestWithBody_ServerDisconnect_AbortStreamsAndTh
...
+                int streamId = await connection.ReadRequestHeaderAsync();
+                await connection.SendDefaultResponseHeadersAsync(streamId);
+                _output.WriteLine("Server sent response headers!");
+                // Server sends GOAWAY frame
+                await connection.SendGoAway(streamId, ProtocolErrors.ENHANCE_YOUR_CALM);
+                _output.WriteLine("Server sent GOAWAY!");
+                connection.ShutdownSend();
+            }).WaitAsync(TimeSpan.FromSeconds(10));
```
Comment (MihaZupan) [inline src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http2.cs:1084]:
```suggestion
            });
```

`CreateClientAndServerAsync` already uses a (much longer) timeout internally.

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs [hunk 4]
```diff
@@ -410,14 +412,10 @@ private async ValueTask<FrameHeader> ReadFrameAsync(bool initialFrame = false)
...
-                            ThrowMissingFrame();
-                        }
-                        else
-                        {
-                            ThrowPrematureEOF(FrameHeader.Size);
-                        }
+                        HttpIOException exception = _incomingBuffer.ActiveLength == 0 ?
+                            GetMissingFrameException() : GetPrematureEOFException(FrameHeader.Size);
+                        throw _goAwayErrorCode is not null ?
+                            new HttpProtocolException((long)_goAwayErrorCode, exception.Message, exception) : exception;
```
Comment (MihaZupan) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:418]:
```suggestion
                        if (_goAwayErrorCode is not null)
                        {
                            ThrowProtocolError(_goAwayErrorCode.Value, SR.net_http_http2_connection_close);
                        }
                        else if (_incomingBuffer.ActiveLength == 0)
                        {
                            ThrowMissingFrame();
                        }
                        else
                        {
                            ThrowPrematureEOF(FrameHeader.Size);
                        }
```

I think we should match the exception message you'd get from a GoAway normally here.

**Review by MihaZupan — APPROVED** [review]
Glad to see this was a simple enough change to get into 9.0.
Thanks

**Comment by liveans** [comment]
> What does the RFC says? I'd expect server is supposed to let us finish those requests.

[RFC9113 - HTTP/2](https://datatracker.ietf.org/doc/html/rfc9113) says:
> Activity on streams numbered lower than or equal to the last stream identifier might still complete successfully. The sender of a GOAWAY frame might gracefully shut down a connection by sending a GOAWAY frame, maintaining the connection in an "open" state until all in-progress streams complete.

If I interpret this correctly, this can go either way.

**Comment by antonfirsov** [comment]
> Not sure if this is the correct way to fix this.

I think @JamesNK should be able to test private binaries to confirm.

**Comment by liveans** [comment]
> > Not sure if this is the correct way to fix this.
> 
> I think @JamesNK should be able to test private binaries to confirm.

That would be nice, but the test is written as he described in the reproduction steps. So, I think this kind of confirms that it's behaving as expected.

**Comment by JamesNK** [comment]
Something that occurred to me, can responseStream.ReadAsync throw `HttpProtocolException` in other scenarios? If reading from the stream only throws `IOException` and friends, should the error by wrapped in `HttpIOException`?

**Comment by liveans** [comment]
> Something that occurred to me, can responseStream.ReadAsync throw `HttpProtocolException` in other scenarios? If reading from the stream only throws `IOException` and friends, should the error by wrapped in `HttpIOException`?

That's a great question, so you're saying that this can be breaking change. I'll leave the answer to the HTTP experts.

**Comment by JamesNK** [comment]
`ReadAsync` throws a protocol error in some situations so it's not a new error. That is what I was worried about.

Is it breaking change? Maybe. I don't know the semantics of when to throw an `IOException` vs `HttpProtocolException`.

**Comment by ManickaP** [comment]
`HttpProtocolException` derives from `HttpIOException` which derives from `IOException`.

**Pull request merged**